adapters/Mintegral/MintegralAdapter/Bidding/GADMAdapterMintegralNativeAdLoader.m

```diff
@@ -0,0 +1,245 @@
+        }
+    }
+}
+
+- (void)nativeAdsFailedToLoadWithError:(nonnull NSError *)error bidNativeManager:(nonnull MTGBidNativeAdManager *)bidNativeManager {
+    if (_adLoadCompletionHandler) {
+        _adLoadCompletionHandler(nil,error);
+    }
+}
+
+- (void)nativeAdDidClick:(nonnull MTGCampaign *)nativeAd bidNativeManager:(nonnull MTGBidNativeAdManager *)bidNativeManager {
+    [_adEventDelegate reportClick];
+}
+
+- (void)nativeAdImpressionWithType:(MTGAdSourceType)type bidNativeManager:(nonnull MTGBidNativeAdManager *)bidNativeManager {
+    [_adEventDelegate reportImpression];
+}
+
+#pragma mark - MTGMediaViewDelegate
+- (void)MTGMediaViewWillEnterFullscreen:(MTGMediaView *)mediaView {
+    [_adEventDelegate willPresentFullScreenView];
+}
+
+- (void)MTGMediaViewDidExitFullscreen:(MTGMediaView *)mediaView {
+    [_adEventDelegate didDismissFullScreenView];
+}
+
+- (void)MTGMediaViewVideoDidStart:(MTGMediaView *)mediaView {
+    [_adEventDelegate didPlayVideo];
+}
+
+- (void)MTGMediaViewVideoPlayCompleted:(MTGMediaView *)mediaView {
+    [_adEventDelegate didEndVideo];
+}
+
+- (void)nativeAdDidClick:(nonnull MTGCampaign *)nativeAd mediaView:(MTGMediaView *)mediaView {
+    [_adEventDelegate reportClick];
+}
+
+- (void)nativeAdImpressionWithType:(MTGAdSourceType)type mediaView:(MTGMediaView *)mediaView {
+    [_adEventDelegate reportImpression];
+}
+
+#pragma mark - GADMediatedUnifiedNativeAd
+- (NSString *)headline {
+    return _campaign.appName;
+}
+
+- (NSArray *)images {
+    return _images;
+}
+
+- (NSString *)body {
+    return _campaign.appDesc;
+}
+
+- (GADNativeAdImage *)icon {
+    return _icon;
+}
+
+- (NSString *)callToAction {
+    return _campaign.adCall;
+}
+
+- (NSDecimalNumber *)starRating {
+    NSString *star = [NSString stringWithFormat:@"%@",[_campaign valueForKey:@"star"]];
+    return [NSDecimalNumber decimalNumberWithString:star];
+}
+
+- (NSString *)store {
+    return nil;
+}
+
+- (NSString *)price {
+    return nil;
+}
+
+-(NSString *)advertiser{
+    return nil;
+}
+
+- (NSDictionary *)extraAssets {
+    return nil;
+}
+
+- (BOOL)hasVideoContent {
+    return YES;
+}
+
+-  (UIView *)mediaView{
+    [_mediaView setMediaSourceWithCampaign:_campaign unitId:_adUnitId];
+    return _mediaView;
+}
+
+#pragma mark - GADMediationNativeAd
+- (BOOL)handlesUserClicks{
+    return YES;
+}
+
+- (BOOL)handlesUserImpressions{
+    return YES;
+}
+
+#pragma mark - GADMediatedUnifiedNativeAd
+-(void)didRenderInView:(UIView *)view clickableAssetViews:(NSDictionary<GADNativeAssetIdentifier,UIView *> *)clickableAssetViews nonclickableAssetViews:(NSDictionary<GADNativeAssetIdentifier,UIView *> *)nonclickableAssetViews viewController:(UIViewController *)viewController{
+    [_nativeManager registerViewForInteraction:view withClickableViews:clickableAssetViews.allValues withCampaign:_campaign];
+}
+
+- (GADNativeAdImage *)loadImageWithURLString:(NSString *)urlString {
+    if (!urlString.length) {
```

Comment:
@joshh-devrel 
`- (GADNativeAdImage *)imageWithUrlString:(NSString *)urlString {
  NSURL *url = [NSURL URLWithString:urlString];
  NSData *data = [NSData dataWithContentsOfURL:url];
  UIImage *image = [UIImage imageWithData:data];
  return [[GADNativeAdImage alloc] initWithImage:image];
}

`
If I use this method to return a GADNativeAdImage object, will dataWithContentsOfURL/imageWithData freeze the main thread?

Reply:
@Mintegral-Lee You can follow a similar implementations as in [this example](https://github.com/googleads/googleads-mobile-ios-mediation/blob/main/adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m#L92-L133), using the [`dataTaskWithURL:completionHandler:`](https://developer.apple.com/documentation/foundation/nsurlsession/1410330-datataskwithurl) method.